src/app/delegation_backend/src/submit.go

```diff
@@ -0,0 +1,202 @@
+	blockHashBytes := blake2b.Sum256(req.Data.Block.data)
+	blockHash := base58.CheckEncode(blockHashBytes[:], BASE58CHECK_VERSION_BLOCK_HASH)
+	createdAtStr := req.Data.CreatedAt.UTC().Format(time.RFC3339)
+	pkStr := base58.CheckEncode(req.Submitter[:], BASE58CHECK_VERSION_PK)
+	res.metaPath = strings.Join([]string{"submissions", pkStr, blockHash, createdAtStr + ".json"}, "/")
+	res.blockPath = "blocks/" + blockHash + ".dat"
+	return
+}
+
+func makeSignPayload(req *submitRequestData) ([]byte, error) {
+	createdAtStr := req.CreatedAt.UTC().Format(time.RFC3339)
+	createdAtJson, err2 := json.Marshal(createdAtStr)
+	if err2 != nil {
+		return nil, err2
+	}
+	signPayload := new(BufferOrError)
+	signPayload.WriteString("{\"block\":")
+	signPayload.Write(req.Block.json)
+	signPayload.WriteString(",\"created_at\":")
+	signPayload.Write(createdAtJson)
+	signPayload.WriteString(",\"peer_id\":")
+	signPayload.Write(req.PeerId.json)
+	if req.SnarkWork != nil {
+		signPayload.WriteString(",\"snark_work\":")
+		signPayload.Write(req.SnarkWork.json)
+	}
+	signPayload.WriteString("}")
+	return signPayload.Buf.Bytes(), signPayload.Err
+}
+
+// TODO consider using pointers and doing `== nil` comparison
+var nilSig Sig
+var nilPk Pk
+var nilTime time.Time
+
+func (h *SubmitH) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.ContentLength == -1 {
+		w.WriteHeader(411)
+		return
+	} else if r.ContentLength > MAX_SUBMIT_PAYLOAD_SIZE {
+		w.WriteHeader(413)
+		return
+	}
+	body, err1 := io.ReadAll(io.LimitReader(r.Body, r.ContentLength))
+	if err1 != nil || int64(len(body)) != r.ContentLength {
+		h.app.Log.Debugf("Error while reading /submit request's body: %v", err1)
+		w.WriteHeader(400)
+		writeErrorResponse(h.app, &w, "Error reading the body")
+		return
+	}
+	var req submitRequest
+	if err := json.Unmarshal(body, &req); err != nil {
+		h.app.Log.Debugf("Error while unmarshaling JSON of /submit request's body: %v", err)
+		w.WriteHeader(400)
+		writeErrorResponse(h.app, &w, "Wrong payload")
+		return
+	}
+	if req.Data.Block == nil || req.Data.PeerId == nil || req.Data.CreatedAt == nilTime || req.Submitter == nilPk || req.Sig == nilSig {
+		h.app.Log.Debug("One of required fields wasn't provided")
+		w.WriteHeader(400)
+		writeErrorResponse(h.app, &w, "One of required fields wasn't provided")
+		return
+	}
+
+	wl := h.app.Whitelist.ReadWhitelist()
+	if (*wl)[req.Submitter] == nil {
+		w.WriteHeader(401)
+		writeErrorResponse(h.app, &w, "Submitter is not registered")
+		return
+	}
+
+	submittedAt := h.app.Now()
+	if req.Data.CreatedAt.Add(TIME_DIFF_DELTA).After(submittedAt) {
+		h.app.Log.Debugf("Field created_at is a timestamp in future: %v", submittedAt)
+		w.WriteHeader(400)
+		writeErrorResponse(h.app, &w, "Field created_at is a timestamp in future")
+		return
+	}
+
+	passesAttemptLimit := h.app.SubmitCounter.RecordAttempt(req.Submitter)
+	if !passesAttemptLimit {
+		w.WriteHeader(429)
+		writeErrorResponse(h.app, &w, "Too many requests per hour")
```

Comment:
This is vulnerable to attack: anybody can flood this endpoint with requests claiming to be from a particular pk, and cause the counter to block their legitimate messages to affect their score.

We should check the signature before doing this, to make sure that the message really is from the pk that it alleges.

Reply:
@mrmr1993 very good catch!